File: Framework/Java/framework/framework/smartkernel/framework/drawing/SmartImage.java
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
package smartkernel.framework.drawing;

import java.awt.*;
import java.awt.image.*;
import java.io.*;

import javax.imageio.*;
import javax.imageio.stream.*;

import smartkernel.framework.io.*;

/**
 * 图像类
 * 
 */
public class SmartImage {

	/**
	 * 获取图像
	 * 
	 * @param filePath，文件路径
	 * @return，结果
	 */
	public static BufferedImage fromFile(String filePath) {
		try {
			return ImageIO.read(new File(filePath));
		} catch (Exception err) {
			throw new RuntimeException(err);
		}
	}

	/**
	 * 保存图像为文件
	 * 
	 * @param image
	 *            ，图像
	 * @param filePath
	 *            ，文件路径
	 */
	public static void saveImage(RenderedImage image, String filePath) {
		try {
			ImageIO.write(image, SmartPath.getExtension(filePath), new File(
					filePath));
		} catch (Exception err) {
			throw new RuntimeException(err);
		}
	}

	/**
	 * 转换为内存图像
	 * 
	 * @param image
	 *            ，图像
	 * @return，结果
	 */
	public static BufferedImage toBufferedImage(Image image) {
		int width = image.getWidth(null);
		int height = image.getHeight(null);

		BufferedImage bufferedImage = new BufferedImage(width, height,
				BufferedImage.TYPE_INT_RGB);
		Graphics g = bufferedImage.getGraphics();
		g.drawImage(image, 0, 0, null);
		g.dispose();

		return bufferedImage;
	}

	/**
	 * 生成缩略图
	 * 
	 * @param sourceFilePath
	 *            ，原图片路径
	 * @param thumbnailFilePath
	 *            ，缩略图保存的路径
	 * @param scale
	 *            ，缩小的尺寸（宽或高的最大边）
	 */
	public static void toThumbnailImage(String sourceFilePath,
			String thumbnailFilePath, int scale) {

		try {
			BufferedImage image = ImageIO.read(new File(sourceFilePath));
			float oldHeight = image.getHeight();
			float oldWidth = image.getWidth();
			int height = 0;
			int width = 0;
			float propotion = 0;
			if (oldHeight > oldWidth) {
				height = scale;
				propotion = oldHeight / scale;
				width = (int) (oldWidth / propotion);
			} else {
				width = scale;
				propotion = oldWidth / scale;
				height = (int) (oldHeight / propotion);
			}

			Image thumbnailImage = image.getScaledInstance(width, height,
					Image.SCALE_SMOOTH);
			saveImage(toBufferedImage(thumbnailImage), thumbnailFilePath);
		} catch (Exception err) {
			throw new RuntimeException(err);
		}
	}

	/**
	 * 生成缩略图
	 * 
	 * @param originalImage
	 *            ，原始图片
	 * @param scale
	 *            ，缩小的尺寸（宽或高的最大边）
	 */
	public static BufferedImage toThumbnailImage(BufferedImage originalImage,
			int scale) {

		try {
			BufferedImage image = originalImage;
			float oldHeight = image.getHeight();
			float oldWidth = image.getWidth();
			int height = 0;
			int width = 0;
			float propotion = 0;
			if (oldHeight > oldWidth) {
				height = scale;
				propotion = oldHeight / scale;
				width = (int) (oldWidth / propotion);
			} else {
				width = scale;
				propotion = oldWidth / scale;
				height = (int) (oldHeight / propotion);
			}

			Image thumbnailImage = image.getScaledInstance(width, height,
					Image.SCALE_SMOOTH);
			return toBufferedImage(thumbnailImage);
		} catch (Exception err) {
			throw new RuntimeException(err);
		}
	}

	/**
	 * 调整图片大小
	 * 
	 * @param originalImage
	 *            ，原始图片
	 * @param newWidth
	 *            ，新的大小
	 * @param newHeight
	 *            ，新的大小
	 * @return，调整之后的图片
	 */
	public static BufferedImage resize(BufferedImage originalImage,
			int newWidth, int newHeight) {
		int originalWidth = originalImage.getWidth();
		int originalHeight = originalImage.getHeight();

		if (originalWidth > originalHeight) {
			newHeight = (int) Math
					.floor(((newWidth * originalHeight) / (double) originalWidth));
		} else if (originalWidth < originalHeight) {
			newWidth = newHeight;
			newHeight = (int) Math
					.floor(((newHeight * originalHeight) / (double) originalWidth));
		} else {
			newHeight = newWidth;
		}

		BufferedImage bufferedImage = new BufferedImage(newWidth, newHeight,
				originalImage.getType());
		Graphics2D g = bufferedImage.createGraphics();
		g.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
				RenderingHints.VALUE_INTERPOLATION_BILINEAR);
		g.drawImage(originalImage, 0, 0, newWidth, newHeight, 0, 0,
				originalWidth, originalHeight, null);
		g.dispose();
		return bufferedImage;
	}

	/**
	 * 修剪图片
	 * 
	 * @param originalImage
	 *            ，原始图片
	 * @param croppedArea
	 *            ，修剪区域
	 * @return，修剪之后的图片
	 */
	public static BufferedImage crop(BufferedImage originalImage,
			Rectangle croppedArea) {
		return originalImage.getSubimage(0, 0, croppedArea.width,
				croppedArea.height);
	}

	/**
	 * 转换图片格式
	 * 
	 * @param originalImage
	 *            ，原始图片
	 * @param format
	 *            ，新格式
	 * @return，转换之后的图片
	 */
	public static BufferedImage toImageFormat(BufferedImage originalImage,
			String format) {
		try {
			ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
			MemoryCacheImageOutputStream memoryCacheImageOutputStream = new MemoryCacheImageOutputStream(
					byteArrayOutputStream);
			ImageIO.write(originalImage, format, memoryCacheImageOutputStream);

			return ImageIO.read(SmartStream.convert(byteArrayOutputStream));
		} catch (Exception err) {
			throw new RuntimeException(err);
		}
	}
}

File: Framework/CSharp/Framework/Framework/Drawing/SmartImage.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.IO;
using System.Linq;
using System.Windows.Media.Imaging;

namespace Smartkernel.Framework.Drawing
{
    /// <summary>
    /// 图像类
    /// </summary>
    public static class SmartImage
    {
        private static bool ThumbnailCallback()
        {
            return true;
        }

        /// <summary>
        /// 生成缩略图
        /// </summary>
        /// <param name="sourceFilePath">原图片路径</param>
        /// <param name="thumbnailFilePath">缩略图保存的路径</param>
        /// <param name="scale">缩小的尺寸（宽或高的最大边）</param>
        public static void ToThumbnailImage(string sourceFilePath, string thumbnailFilePath, int scale)
        {
            using (var image = Image.FromFile(sourceFilePath))
            {
                float oldHeight = image.Height;
                float oldWidth = image.Width;
                var height = 0;
                var width = 0;
                float propotion = 0;
                if (oldHeight > oldWidth)
                {
                    height = scale;
                    propotion = oldHeight / scale;
                    width = Convert.ToInt16(oldWidth / propotion);
                }
                else
                {
                    width = scale;
                    propotion = oldWidth / scale;
                    height = Convert.ToInt16(oldHeight / propotion);
                }

                using (var thumbnailImage = image.GetThumbnailImage(width, height, ThumbnailCallback, IntPtr.Zero))
                {
                    using (var output = new Bitmap(thumbnailImage))
                    {
                        output.Save(thumbnailFilePath, ImageFormat.Jpeg);
                    }
                }
            }
        }

        /// <summary>
        /// 调整图片大小
        /// </summary>
        /// <param name="originalImage">原始图片</param>
        /// <param name="newSize">新的大小</param>
        /// <param name="disposeOriginal">是否释放原来的图片资源</param>
        /// <returns>调整之后的图片</returns>
        public static Bitmap Resize(Bitmap originalImage, Size newSize, bool disposeOriginal = true)
        {
            Bitmap resizedImage = null;

            var pixelFormats = new[] { PixelFormat.Format1bppIndexed, PixelFormat.Format1bppIndexed, PixelFormat.Format4bppIndexed, PixelFormat.Format8bppIndexed, PixelFormat.Undefined, PixelFormat.DontCare, PixelFormat.Format16bppArgb1555, PixelFormat.Format16bppGrayScale };

            if (!pixelFormats.ToList().Contains(originalImage.PixelFormat))
            {
                if (originalImage.Width > originalImage.Height)
                {
                    resizedImage = new Bitmap(newSize.Width, (int)Math.Floor((newSize.Width * originalImage.Height) / (double)originalImage.Width), originalImage.PixelFormat);
                }
                else if (originalImage.Width < originalImage.Height)
                {
                    resizedImage = new Bitmap(newSize.Height, (int)Math.Floor((newSize.Height * originalImage.Height) / (double)originalImage.Width), originalImage.PixelFormat);
                }
                else
                {
                    resizedImage = new Bitmap(newSize.Width, newSize.Width, originalImage.PixelFormat);
                }
                resizedImage.SetResolution(originalImage.HorizontalResolution, originalImage.VerticalResolution);

                using (var resizedImageGraphic = Graphics.FromImage(resizedImage))
                {
                    resizedImageGraphic.CompositingQuality = CompositingQuality.HighQuality;
                    resizedImageGraphic.SmoothingMode = SmoothingMode.HighQuality;
                    resizedImageGraphic.InterpolationMode = InterpolationMode.HighQualityBicubic;
                    resizedImageGraphic.DrawImage(originalImage, 0, 0, resizedImage.Width, resizedImage.Height);
                }

                if (disposeOriginal)
                {
                    originalImage.Dispose();
                }
            }

            return resizedImage;
        }

        /// <summary>
        /// 修剪图片
        /// </summary>
        /// <param name="originalImage">原始图片</param>
        /// <param name="croppedArea">修剪区域</param>
        /// <param name="disposeOriginal">是否释放资源</param>
        /// <returns>修剪之后的图片</returns>
        public static Bitmap Crop(Bitmap originalImage, Rectangle croppedArea, bool disposeOriginal = true)
        {
            var croppedImage = new Bitmap(croppedArea.Width, croppedArea.Height, originalImage.PixelFormat);
            croppedImage.SetResolution(originalImage.HorizontalResolution, originalImage.VerticalResolution);
            using (var croppedImageGraphic = Graphics.FromImage(croppedImage))
            {
                croppedImageGraphic.DrawImage(originalImage, 0, 0, croppedArea, GraphicsUnit.Pixel);
            }

            if (disposeOriginal)
            {
                originalImage.Dispose();
            }

            return croppedImage;
        }

        /// <summary>
        /// 转换图片格式
        /// </summary>
        /// <param name="originalImage">原始图片</param>
        /// <param name="format">新格式</param>
        /// <param name="disposeOriginal">是否释放原始图片的资源</param>
        /// <returns>转换之后的图片</returns>
        public static Bitmap ToImageFormat(Bitmap originalImage, ImageFormat format, bool disposeOriginal = true)
        {
            Bitmap newImage = null;
            var newImageStream = new MemoryStream();
            originalImage.Save(newImageStream, format);
            newImage = new Bitmap(newImageStream);

            if (disposeOriginal)
            {
                originalImage.Dispose();
            }

            return newImage;
        }

        /// <summary>
        /// 生成缩略图
        /// </summary>
        /// <param name="originalImage">原始图片</param>
        /// <param name="scale">缩小的尺寸（宽或高的最大边）</param>
        public static Bitmap ToThumbnailImage(Bitmap originalImage, int scale)
        {
            float oldHeight = originalImage.Height;
            float oldWidth = originalImage.Width;
            var height = 0;
            var width = 0;
            float propotion = 0;
            if (oldHeight > oldWidth)
            {
                height = scale;
                propotion = oldHeight / scale;
                width = Convert.ToInt16(oldWidth / propotion);
            }
            else
            {
                width = scale;
                propotion = oldWidth / scale;
                height = Convert.ToInt16(oldHeight / propotion);
            }

            return new Bitmap(originalImage.GetThumbnailImage(width, height, ThumbnailCallback, IntPtr.Zero));
        }

        /// <summary>
        /// 转换为BitmapImage格式
        /// </summary>
        /// <param name="originalImage">原始图像</param>
        /// <returns>结果</returns>
        public static BitmapImage ToBitmapImage(Bitmap originalImage)
        {
            //这里不能释放MemoryStream，否则无法获得BitmapImage
            MemoryStream memoryStream = new MemoryStream();
            originalImage.Save(memoryStream, ImageFormat.Png);
            memoryStream.Position = 0;
            BitmapImage bitmapImage = new BitmapImage();
            bitmapImage.BeginInit();
            bitmapImage.StreamSource = memoryStream;
            bitmapImage.EndInit();
            return bitmapImage;
        }
    }
}
